### DISCUSSION TITLE
How to generate white matter tractography?

### DISCUSSION THREAD

**Discussion by sunitaiitb** [dipy/dipy#3068] [Tractography / Tractometry]
Dear DIPY contributors,

We are trying to generate tractography map of brain. DIPY is the library used for this. The instruction we are getting is very confusing. Sometimes DTI is used and sometimes DWI is used. We could understand that .bvac and .bval files generated from DWI image is used for the generation of fiber tracts.

From the web, we find that DTI should have four dimensions however, the DTI file we are having has only three dimensions. How to add the fourth dimension in DTI FA image which provide information about the magnitude and direction of the diffusion?

Which method is correct for the generation of white matter mask "median_otsu" or "tissue_classification"?

Could you please guide us on how to generate the tractography image.

We will highly appreciate your help. 

Thank you so much.

Best regards,
Sunita

**Comment by pjsjongsung** [comment]
It seems you are looking at the FA image, which is a processed image from the DTI, thus being a 3 dimensional image. There must be some volume you feed to your function creating a FA image.

For generating white matter mask, if it is not a skull stripped image, you want to do median_otsu first and then tissue_classification. If the brain is already extracted, you can just to tissue classification.

We have steps generating tracts in our [tutorials](https://docs.dipy.org/stable/examples_built/index#fiber-tracking). Hopefully they would be helpful, if not let us know.

**Reply by sunitaiitb** [reply]
Which image is required for tractography, is it DTI or DWI? The tutorial given is working fine where the data used has four dimensions.  In the tutorial it is not clear which image is used.  The DTI files we are having has three dimensions.  In the tutorial the file used has four dimensions. Do we need to add the fourth dimension into the existing DTI file?  If yes then which DTI information should we add. Sorry we are asking beginner's questions.  

We are using T1-Weighted for tissue classification. We have .bvac and .bval files. 

Label 1 represents white matter. Am I right?
Is the label 2 used for initiating tracking?

 We are stuck in the first step itself.  Your guidance will help us to solve this problem. 

Thank you so much. 

Best regards,
Sunita

**Reply by skoudoro** [reply]
Hi @sunitaiitb,

To do a tractography, you need a tensor image.

the DTI algorithm transforms your original data (DWI) to a tensor image(DTI).

The tractography algorithms will use the Tensor  image (DTI) to generates peak and do the full tracking.

The original dimension (4D DWI image) is used to create your tensor image (DTI).

The mask created with T1 or FA image is just to help the tractography. Typically, the mask is composed by 0 or 1.
if zero, do not track this area if 1, you can. Thresholding FA or tissue classification of T1 help you to create this mask containing 0 and 1.

> Label 1 represents white matter. Am I right?
> Is the label 2 used for initiating tracking?

we do not know, depends of your convention and what you put in these label. You will have to display then with a software to see what each label represent in your case.

I recommend you to attend the DIPY Workshop 2024 in March, which will explain you in details all the concepts. You will be able to ask all the questions you want and many experts will be able to help you with your code during some specific session: 

see here: https://dipy.org/workshop/latest

I hope it helps

**Reply by pjsjongsung** [reply]
Also if you are using T1 for tissue classification, consider using `dipy_evac_plus` command or `dipy.nn.evac.EVACPlus`  in code to get a brain mask first and then do tissue classification following the [tutorials](https://docs.dipy.org/stable/examples_built/segmentation/tissue_classification.html#sphx-glr-examples-built-segmentation-tissue-classification-py).

**Reply by sunitaiitb** [reply]
Dear Jong and Serge,

Thank you so much for your response. 

The problem we are facing is that we are not having the original dataset. 
Our data is having T1W, DWI, .bvac and .bval files.  The dimensions are as follows. 

T1W: 240, 240, 155
DWI: 240, 240, 155 (Instead of 4 dimensions we are having only 3 dimensions)
BVAC: 56,3
BVAL: 56

Is it possible to append these vector information into DWI because the tutorial shared by DIPY takes 4D to generate tractography image?

Your help will be highly appreciated to understand and build tractography image using DIPY libraries which is widely used for the purpose. 
 
Thanks again for your guidance. 

Best regards,
Sunita

**Reply by pjsjongsung** [reply]
It is strange that you have only 3 dimensions in DWI. You might be looking at a processed image, such as FA. bval and bvec is information about the gradient of each DWI volume, and does not contain information about the signal intensities. So it seems without the original dataset with the 4 volumes, it is not possible to do tractography.

**Reply by sunitaiitb** [reply]
Dear Jong,

The dataset is from TCIA site which provides incomplete information.  For a beginner it is really confusing. 

Finally, we got a dataset with DWI in 4D and T1W in 3D.  The dimension of DWI is 128, 128, 88, 67  and T1W is 176, 240, 256.  The WM mask should match with DWI dimension. For that do we have to coregister and resample T1W to match 128, 128, 88 dimension before extracting WM mask?

or

Is it possible to generate WM mask from DWI image?
We used median_otsu to get the skull stripped brain mask then used DWI image and mask in dti.tensormodel to get FA image.
Now can we get the WM mask from FA image?

Thank you so much. 

Best regards,
Sunita

**Reply by pjsjongsung** [reply]
While it is best if you have a calculate WM mask, it is also possible to get a crude mask by thresholding the FA image. This [tutorial](https://docs.dipy.org/stable/examples_built/fiber_tracking/tracking_stopping_criterion.html#sphx-glr-examples-built-fiber-tracking-tracking-stopping-criterion-py) could provide some information on stopping criterions.

**Reply by sunitaiitb** [reply]
Dear Jong,

Thanks for your guidance. 
At least now we could generate the tracts. But to generate meaningful tracts we need understand the functions.  The tractograph looks weird may be because of wrong selection of seed. 

We loaded the DWI data with bvec and bval values as follows.
```
data, affine, img = load_nifti(dwi_, return_img=True)
bvals, bvecs = read_bvals_bvecs(val_, vec_)
gtab = gradient_table(bvals, bvecs) 
```
Then we ran median_otsu for brain extraction of DWI.
The shape of DWI is 128,128,88,67.  What does vol_idx[3] mean here? 

`dwi_data, dwi_mask = median_otsu(data, vol_idx=[3])`


We used TensorModel() to generate the FA. The mask we got from median_ostu() is in boolean form. 
```
dti_model = TensorModel(gtab)
dti_fit = dti_model.fit(dwi_data, mask=dwi_mask)
FA = fractional_anisotropy(dti_fit.evals)
```

When we visualize with 0.2 threshold we got following image. Do you think this image is acceptable for serving as white matter (WM) mask?
![White_Matter_with_thereshold_0 2](https://github.com/dipy/dipy/assets/94822781/7340447e-2118-469d-a145-d3892323a43e)


 We are taking FA as white matter. Now how to set the seed mask? We tried setting greater than 0.6 threshold as the seed mask. With this threshold the image appeared as follows.   
![threshold_gt_0 6_as_seed](https://github.com/dipy/dipy/assets/94822781/bd39091c-cc26-4127-8323-31b2d34e78df)


```
while_matter=FA.copy()
seed_mask = (labels == 2)
seed_mask[white_matter < 0.5] = 0
seeds = utils.seeds_from_mask(seed_mask, affine, density=2)

``` 

After setting the seed we ran the following code. But it ran for 50% and stopped. Is there anything to change here?

```
response, ratio = auto_response_ssst(gtab, data, roi_radii=10, fa_thr=0.7)
csd_model = ConstrainedSphericalDeconvModel(gtab, response)
csd_fit = csd_model.fit(data, mask=white_matter)

dg = DeterministicMaximumDirectionGetter.from_shcoeff(csd_fit.shm_coeff,
                                                      max_angle=30.,
                                                      sphere=default_sphere) 

```
After that we ran the local tracking and visualize the tractograph.

```
interactive=True
streamline_generator = LocalTracking(dg,
                                     stopping_criterion,
                                     seeds,
                                     affine,
                                     step_size=.5,
                                     return_all=True)
streamlines = Streamlines(streamline_generator)
sft = StatefulTractogram(streamlines, img, Space.RASMM)
if has_fury:
    scene = window.Scene()
    scene.add(actor.line(streamlines, colormap.line_colors(streamlines)))
    window.record(scene, out_path='tractogram_deterministic_thresh_all.png',
                  size=(800, 800))
    if interactive:
        window.show(scene)
```
After this we are getting this image. 
![tracts_with_threshold_0 6_as_seed](https://github.com/dipy/dipy/assets/94822781/cf3d645d-4050-4347-9833-5fe409d1d083)



In this image the threshold values greater than 0.6 is used as seed for generation of fiber tracts.  How to decide the seed? How to get the correct tractograph? How to generate corticospinal tract, corona radiata, corpus collasum etc?

Thanks a lot.
Sunita

**Reply by Sunita2023** [reply]
Dear Jong,

We have understood on how to generate white matter tracts. Thanks DIPY team for the wonderful tutorials.  We generated the full brain streamlines and extracted the bundle of interest by comparing the recobundles.  

Thank you all DIPY developers. 

A little more guidance from DIPY team would have been a great help.  
![fornix_centroids](https://github.com/dipy/dipy/assets/112482239/53016c3b-faf4-4faf-ac43-dd5d6c354cd5)

Kindly please provide some guidance when we asked question.   

Thank you so much.
Sunita

**Reply by pjsjongsung** [reply]
Glad it worked! :) I'm sorry I missed the previous question. Please feel free to add a comment or ping me as a reminder if there is no response in the future.
